In [1]:
from pathlib import Path
import math

import numpy as np
import pandas as pd
from sklearn import metrics

from sklearn.ensemble import RandomForestRegressor
from catboost import CatBoostRegressor, CatBoostClassifier, Pool, cv
from sklearn.model_selection import cross_val_score, KFold, GridSearchCV

In [2]:
from eclyon.transforms import process_df, split_vals

In [3]:
path_to_repo = Path('..').resolve()
path_to_data = path_to_repo / 'data'

In [4]:
df= pd.read_csv(path_to_data / 'clean_health.csv')
df

gender  age  height  weight  waistline  sight_left  sight_right  \
0         True   35     170      75       90.0         1.0          1.0   
1         True   30     180      80       89.0         0.9          1.2   
2         True   40     165      75       91.0         1.2          1.5   
3         True   50     175      80       91.0         1.5          1.2   
4         True   50     165      60       80.0         1.0          1.2   
...        ...  ...     ...     ...        ...         ...          ...   
991341    True   45     175      80       92.1         1.5          1.5   
991342    True   35     170      75       86.0         1.0          1.5   
991343   False   40     155      50       68.0         1.0          0.7   
991344    True   25     175      60       72.0         1.5          1.0   
991345    True   50     160      70       90.5         1.0          1.5   

        hear_left  hear_right    SBP  ...  LDL_chole  triglyceride  \
0             1.0         1.0  120.0  ...      126.0          92.0   
1             1.0         1.0  130.0  ...      148.0         121.0   
2             1.0         1.0  120.0  ...       74.0         104.0   
3             1.0         1.0  145.0  ...      104.0         106.0   
4             1.0         1.0  138.0  ...      117.0         104.0   
...           ...         ...    ...  ...        ...           ...   
991341        1.0         1.0  114.0  ...      125.0         132.0   
991342        1.0         1.0  119.0  ...       84.0          45.0   
991343        1.0         1.0  110.0  ...       77.0         157.0   
991344        1.0         1.0  119.0  ...       73.0          53.0   
991345        1.0         1.0  133.0  ...      153.0         163.0   

        hemoglobin  urine_protein  serum_creatinine  SGOT_AST  SGOT_ALT  \
0             17.1            1.0               1.0      21.0      35.0   
1             15.8            1.0               0.9      20.0      36.0   
2             15.8            1.0               0.9      47.0      32.0   
3             17.6            1.0               1.1      29.0      34.0   
4             13.8            1.0               0.8      19.0      12.0   
...            ...            ...               ...       ...       ...   
991341        15.0            1.0               1.0      26.0      36.0   
991342        15.8            1.0               1.1      14.0      17.0   
991343        14.3            1.0               0.8      30.0      27.0   
991344        14.5            1.0               0.8      21.0      14.0   
991345        15.8            1.0               0.9      24.0      43.0   

        gamma_GTP  smoker  drinker  
0            40.0       1        1  
1            27.0       3        0  
2            68.0       1        0  
3            18.0       1        0  
4            25.0       1        0  
...           ...     ...      ...  
991341       27.0       1        0  
991342       15.0       1        0  
991343       17.0       3        1  
991344       17.0       1        0  
991345       36.0       3        1  

[991346 rows x 24 columns]

In [5]:
df, y, nas = process_df(df,'drinker')

In [6]:
print(len(df))

991346


We will use around 90% of the dataset as training data and 10% for validation

In [7]:
n_total = len(df)
n_valid = 12000
n_train = n_total - n_valid
n_small = 20000

X_train, X_valid = split_vals(df, n_train)
y_train, y_valid = split_vals(y, n_train)

X_small, _ = split_vals(df, n_small)
y_small, _ = split_vals(y, n_small)

y_train = y_train.tolist()
y_valid = y_valid.tolist()
y_small = y_small.tolist()

print('Number of small training data points: X = {}, y = {}'.format(X_small.shape, len(y_small)))
print('Number of full training data points: X = {}, y = {}'.format(X_train.shape, len(y_train)))
print('Number of validation data points: X = {}, y = {}'.format(X_valid.shape, len(y_valid)))

Number of small training data points: X = (20000, 23), y = 20000
Number of full training data points: X = (979346, 23), y = 979346
Number of validation data points: X = (12000, 23), y = 12000


I will use CatBoostClassifier to fit my model. First testing it with my small dataset

In [8]:
model_small = CatBoostClassifier(
    n_estimators = 5000,   # iterations, n_estimators, num_boost_round, num_trees
    learning_rate = 0.001, # learning_rate, eta,
    random_seed = 42,
    allow_writing_files = False,
    logging_level = 'Silent',
)

In [9]:
model_small.fit(X_small, y_small)
print(model_small.score(X_valid, y_valid)) # classification accuracy on small set

0.73025


In [47]:
param_grid = [
    {'learning_rate': [0.01],
     'depth'         : [4,5,6,7],
    'iterations'    : [2000]
    }]

    
model = CatBoostClassifier(
    random_state = 42
)

# I cannot run the whole param grid because my computer crashes, I change up the values to see various results 

In [48]:
tuned_model = GridSearchCV(
    estimator = model, 
    param_grid = param_grid, 
    scoring = None, # uses estimator's default score method 
    refit = True, # keep a fitted version of the overall best model
    cv = 2
)

In [49]:
tuned_model.fit(X_small, y_small);

0:	learn: 0.6903526	total: 4.49ms	remaining: 8.97s
1:	learn: 0.6876574	total: 7.42ms	remaining: 7.42s
2:	learn: 0.6849228	total: 10.8ms	remaining: 7.19s
3:	learn: 0.6824605	total: 14.3ms	remaining: 7.14s
4:	learn: 0.6800234	total: 17.3ms	remaining: 6.91s
5:	learn: 0.6775854	total: 20.1ms	remaining: 6.69s
6:	learn: 0.6753237	total: 23.2ms	remaining: 6.61s
7:	learn: 0.6730307	total: 25.9ms	remaining: 6.46s
8:	learn: 0.6708590	total: 28.8ms	remaining: 6.38s
9:	learn: 0.6684982	total: 31.8ms	remaining: 6.32s
10:	learn: 0.6663187	total: 34.5ms	remaining: 6.23s
11:	learn: 0.6640675	total: 36.8ms	remaining: 6.1s
12:	learn: 0.6620131	total: 39.2ms	remaining: 5.99s
13:	learn: 0.6597612	total: 41.4ms	remaining: 5.87s
14:	learn: 0.6578454	total: 43.6ms	remaining: 5.77s
15:	learn: 0.6557121	total: 45.8ms	remaining: 5.68s
16:	learn: 0.6537707	total: 48.1ms	remaining: 5.61s
17:	learn: 0.6518781	total: 50.4ms	remaining: 5.55s
18:	learn: 0.6499376	total: 52.3ms	remaining: 5.45s
19:	learn: 0.6480827	to

In [50]:
print(tuned_model.best_params_)
print(tuned_model.best_score_)

{'depth': 4, 'iterations': 2000, 'learning_rate': 0.01}
0.7324999999999999


{'depth': 3, 'iterations': 1000, 'learning_rate': 0.01}

score = 0.7313000000000001

{'depth': 3, 'iterations': 2000, 'learning_rate': 0.01}

score = 0.7321

{'depth': 4, 'iterations': 2000, 'learning_rate': 0.01}

score = 0.7324999999999999

In [32]:
def specificity_score(y_true, y_predict):
    '''
    Computes the proba that a label is 0 knowing that the prediction is 0
    '''
    y_predict_0  = [(pred, true) for pred, true in zip(y_predict, y_true) if true == 0]
    y_predict_00 = [(pred, true) for pred, true in y_predict_0 if pred == 0]
    specificity  = (0 if len(y_predict_0) == 0 else len(y_predict_00)/len(y_predict_0))
    return specificity

In [33]:
y_predict_small = model_small.predict(X_valid)

acc = metrics.accuracy_score(y_valid, y_predict_small)
rec = metrics.recall_score(y_valid, y_predict_small)
prc = metrics.precision_score(y_valid, y_predict_small)
spe = specificity_score(y_valid, y_predict_small)
f1  = metrics.f1_score(y_valid, y_predict_small)

print('Accuracy : {:.2f}%'.format(acc*100))
print('Recall : {:.2f}%'.format(rec*100))
print('Precision : {:.2f}%'.format(prc*100))
print('Specificity : {:.2f}%'.format(spe*100))
print('F1-score : {:.2f}%'.format(f1*100))

Accuracy : 73.02%
Recall : 73.44%
Precision : 72.97%
Specificity : 72.60%
F1-score : 73.21%


In [34]:
y_predict_small = tuned_model.predict(X_valid)

acc = metrics.accuracy_score(y_valid, y_predict_small)
rec = metrics.recall_score(y_valid, y_predict_small)
prc = metrics.precision_score(y_valid, y_predict_small)
spe = specificity_score(y_valid, y_predict_small)
f1  = metrics.f1_score(y_valid, y_predict_small)

print('Accuracy : {:.2f}%'.format(acc*100))
print('Recall : {:.2f}%'.format(rec*100))
print('Precision : {:.2f}%'.format(prc*100))
print('Specificity : {:.2f}%'.format(spe*100))
print('F1-score : {:.2f}%'.format(f1*100))

Accuracy : 73.56%
Recall : 73.86%
Precision : 73.55%
Specificity : 73.26%
F1-score : 73.71%


In [56]:
y_hat_boost = tuned_model.predict(X_valid)
confusion = metrics.confusion_matrix(y_hat_boost, y_valid)
print("confusion table of tuned model")
print(confusion)


confusion table of tuned model
[[4380 1574]
 [1599 4447]]


In [57]:
model_full = CatBoostClassifier(
    n_estimators = 2000,   # iterations, n_estimators, num_boost_round, num_trees
    learning_rate = 0.01,# learning_rate, eta,
    depth = 4,
    random_seed = 42,
    allow_writing_files = False,
    logging_level = 'Silent',
)

In [58]:
model_full.fit(X_train, y_train)
print(model_full.score(X_valid, y_valid)) # classification accuracy on full set

0.7386666666666667


In [59]:
y_predict_valid = model_full.predict(X_valid)

acc = metrics.accuracy_score(y_valid, y_predict_valid)
rec = metrics.recall_score(y_valid, y_predict_valid)
prc = metrics.precision_score(y_valid, y_predict_valid)
spe = specificity_score(y_valid, y_predict_valid)
f1  = metrics.f1_score(y_valid, y_predict_valid)

print('Accuracy : {:.2f}%'.format(acc*100))
print('Recall : {:.2f}%'.format(rec*100))
print('Precision : {:.2f}%'.format(prc*100))
print('Specificity : {:.2f}%'.format(spe*100))
print('F1-score : {:.2f}%'.format(f1*100))

Accuracy : 73.87%
Recall : 74.51%
Precision : 73.70%
Specificity : 73.22%
F1-score : 74.10%


In [60]:
model_full.save_model(path_to_repo / 'models' / "model_tuned",
           format="cbm",
           export_parameters=None,
           pool=None)